File: tests/r_parity/43_heckman.R
# StatsPAI Heckman selection parity (R side) -- Module 43.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) dirname(normalizePath(sub("^--file=", "", .file_arg[1]))) else getwd()
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({ library(sampleSelection) })

MODULE <- "43_heckman"
df <- read_csv_strict(MODULE)

# 2-step Heckman matching sp.heckman; sampleSelection::heckit
fit <- sampleSelection::heckit(
  selection = sel ~ z,
  outcome   = y ~ x,
  data      = df,
  method    = "2step"
)

sm <- summary(fit)
co_out <- sm$estimate    # all coefs (selection + outcome)
# Outcome equation rows: "y/(Intercept)", "y/x", "y/invMillsRatio" (or similar)
# Use stat name to pull them; sampleSelection labels them with prefixes.
rn <- rownames(co_out)
get_row <- function(pattern) {
  hit <- grep(pattern, rn, value = TRUE)
  if (length(hit) == 0) return(c(NA, NA))
  v <- co_out[hit[1], ]
  c(v["Estimate"], v["Std. error"])
}

ic <- get_row("(Intercept)$|^o.*Intercept")
xc <- get_row("^o.*x$|^x$|^.*x_")
# Use the outcome equation's variables only; sampleSelection prefixes with
# "out" or has them in a separate slot; safer: directly extract from the
# fit object.

beta_int <- coef(fit, part = "outcome")[["(Intercept)"]]
beta_x   <- coef(fit, part = "outcome")[["x"]]
se_int   <- sqrt(diag(vcov(fit, part = "outcome")))[["(Intercept)"]]
se_x     <- sqrt(diag(vcov(fit, part = "outcome")))[["x"]]

# IMR (lambda) coefficient
lam_est <- coef(fit, part = "outcome")[["invMillsRatio"]]
lam_se  <- sqrt(diag(vcov(fit, part = "outcome")))[["invMillsRatio"]]

rows <- list(
  parity_row(MODULE, "beta_intercept", estimate = beta_int, se = se_int, n = nrow(df)),
  parity_row(MODULE, "beta_x",         estimate = beta_x,   se = se_x,   n = nrow(df)),
  parity_row(MODULE, "lambda_imr",     estimate = lam_est,  se = lam_se, n = nrow(df))
)

write_results(MODULE, rows, extra = list(method = "2-step Heckit",
                                          package = "sampleSelection::heckit"))

File: tests/r_parity/43_heckman.py
"""StatsPAI Heckman selection parity (Python side) -- Module 43."""
from __future__ import annotations
import numpy as np, pandas as pd, statspai as sp
from _common import PARITY_SEED, ParityRecord, dump_csv, write_results

MODULE = "43_heckman"

def make_data(n=600, seed=PARITY_SEED):
    rng = np.random.default_rng(seed)
    x = rng.normal(0, 1, n)
    z = rng.normal(0, 1, n)
    # Selection eqn: sel = 1{0.3 + 0.5*z + u > 0}
    u = rng.normal(0, 1, n)
    sel_lat = 0.3 + 0.5 * z + u
    sel = (sel_lat > 0).astype(int)
    # Outcome eqn (when selected)
    eps = rng.normal(0, 1, n)
    y_lat = 1.0 + 0.5 * x + eps
    y = np.where(sel == 1, y_lat, np.nan)
    return pd.DataFrame({"y": y, "x": x, "z": z, "sel": sel})

def main():
    df = make_data()
    dump_csv(df, MODULE)
    res = sp.heckman(data=df, y="y", x=["x"], select="sel", z=["z"])

    coef = res.detail["coefficient"]
    se = res.detail["se"]
    var_ = res.detail["variable"]
    dct = {str(v): (float(coef[i]), float(se[i])) for i, v in enumerate(var_)}

    rows = []
    for name in ["const", "x"]:
        if name in dct:
            est, sev = dct[name]
            label = "intercept" if name == "const" else name
            rows.append(ParityRecord(MODULE, "py",
                f"beta_{label}", estimate=est, se=sev, n=int(len(df))))
    # lambda (IMR) coefficient
    if "lambda (IMR)" in dct:
        est, sev = dct["lambda (IMR)"]
        rows.append(ParityRecord(MODULE, "py", "lambda_imr",
            estimate=est, se=sev, n=int(len(df))))

    write_results(MODULE, "py", rows,
        extra={"method": "2-step Heckman", "engine": "statspai"})

if __name__ == "__main__":
    main()
